src/topology/continuous_function/zero_at_infty.lean

```diff
@@ -0,0 +1,318 @@
+/-
+Copyright (c) 2022 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import topology.continuous_function.bounded
+
+/-!
+# Continuous functions vanishing at infinity
+
+The type of continuous functions vanishing at infinity. When the domain is compact
+`C(α, β) ≃ (α →C₀ β)` via the identity map. When the codomain is a metric space, every continuous
+map which vanishes at infinity is a bounded continuous function. When the domain is a locally
+compact space, this type has nice properties.
+
+## TODO
+
+* Create more intances of algebraic structures (e.g., `non_unital_semiring`) once the necessary
+  type classes (e.g., `topological_ring`) are sufficiently generalized.
+* Relate the unitization of `α →C₀ β` to the Alexandroff compactification.
+-/
+universes u v w
+
+variables {F : Type*} {α : Type u} {β : Type v} {γ : Type w} [topological_space α]
+
+open_locale bounded_continuous_function topological_space
+open filter metric
+
+/-- `α →C₀ β)` is the type of continuous functions `α → β` which vanish at infinity from a
+topological space to a metric space with a zero element.
+
+When possible, instead of parametrizing results over `(f : α →C₀ β)`,
+you should parametrize over `(F : Type*) [zero_at_infty_continuous_map_class F α β] (f : F)`.
+
+When you extend this structure, make sure to extend `zero_at_infty_continuous_map_class`. -/
+structure zero_at_infty_continuous_map (α : Type u) (β : Type v)
+  [topological_space α] [has_zero β] [topological_space β] extends continuous_map α β :
+  Type (max u v) :=
+(zero_at_infty' : tendsto to_fun (cocompact α) (𝓝 0))
+
+localized "notation  α ` →C₀ ` β := zero_at_infty_continuous_map α β" in zero_at_infty
+
+/-- `zero_at_infty_continuous_map_class F α β` states that `F` is a type of continuous maps which
+vanish at infinity.
+
+You should also extend this typeclass when you extend `zero_at_infty_continuous_map`. -/
+class zero_at_infty_continuous_map_class (F α β : Type*) [topological_space α] [has_zero β]
+  [topological_space β] extends continuous_map_class F α β :=
+(zero_at_infty (f : F) : tendsto f (cocompact α) (𝓝 0))
+
+export zero_at_infty_continuous_map_class (zero_at_infty)
+
+namespace zero_at_infty_continuous_map
+
+section basics
+
+variables [topological_space β] [has_zero β] [zero_at_infty_continuous_map_class F α β]
+
+instance : zero_at_infty_continuous_map_class (α →C₀ β) α β :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by { obtain ⟨⟨_, _⟩, _⟩ := f, obtain ⟨⟨_, _⟩, _⟩ := g, congr' },
+  map_continuous := λ f, f.continuous_to_fun,
+  zero_at_infty := λ f, f.zero_at_infty' }
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`
+directly. -/
+instance : has_coe_to_fun (α →C₀ β) (λ _, α → β) := fun_like.has_coe_to_fun
```

Comment:
@Vierkantor Is this example of `fun_like` inference issues something you would expect?

Reply:
Yes: I think it got better after I made some tweaks to elaboration and `simp` but it was needed in some places for certain morphisms.